features/mbedtls/targets/TARGET_STM/md5_alt.c

```diff
@@ -0,0 +1,158 @@
+/*
+ *  MD5 hw acceleration
+ *******************************************************************************
+ * Copyright (c) 2017, STMicroelectronics
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+#if defined(MBEDTLS_MD5_C)
+#include "mbedtls/md5.h"
+
+#if defined(MBEDTLS_MD5_ALT)
+#include "mbedtls/platform.h"
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = v; while( n-- ) *p++ = 0;
+}
+
+static void st_md5_restore_hw_context(mbedtls_md5_context *ctx)
+{
+    uint32_t i;
+    /* allow multi-instance of HASH use: save context for HASH HW module CR */
+    HASH->STR = ctx->ctx_save_str;
+    HASH->CR = (ctx->ctx_save_cr | HASH_CR_INIT);
+    for (i=0;i<38;i++) {
+        HASH->CSR[i] = ctx->ctx_save_csr[i];
+    }
+}
+
+static void st_md5_save_hw_context(mbedtls_md5_context *ctx)
+{
+    uint32_t i;
+    /* allow multi-instance of HASH use: restore context for HASH HW module CR */
+    ctx->ctx_save_cr = HASH->CR;
+    ctx->ctx_save_str = HASH->STR;
+    for (i=0;i<38;i++) {
+        ctx->ctx_save_csr[i] = HASH->CSR[i];
+    }
```

Comment:
Hello @andresag01 
Here is the full content of the Reference Manual for the context swapping chapter:
Procedure where the data are loaded by software

> The context can be saved only when no block is currently being processed. That is, you
> must wait for DINIS = 1 (the last block has been processed and the input FIFO is empty) or
> NBW ≠ 0 (the FIFO is not full and no processing is ongoing).
> • Context saving:
> Store the contents of the following registers into memory:
> – **HASH_IMR**
> – **HASH_STR**
> – **HASH_CR**
> – HASH_CSR0 to HASH_CSR50 on STM32F415/417xx, and HASH_CSR0 to
> HASH_CSR53 on STM32F43xxx.
> • Context restoring:
> [...]

and a note: 
> Note: If context swapping does not involve HMAC operations, the HASH_CSR38 to
> HASH_CSR50 (STM32F415/417xx) and HASH_CSR38 to HASH_CSR53 (STM32F43xxx)
> registers do not have to be saved and restored.

We are not using the interrupts in mbed-os: I did not save HASH_IMR register. But STR and CR are required to be saved-restored.

I don't save `HASH_CSR38` to `HASH_CSR53 `because we don't use the hardware acceleration for HMAC (it is done by SW at mbedtls level)